pkg/streamingpromql/query.go

```diff
@@ -473,6 +535,26 @@ func (q *Query) populateMatrixFromRangeVectorOperator(ctx context.Context, o typ
 	return m, nil
 }
 
+func (q *Query) populateMatrixFromScalarOperator(d types.ScalarData) promql.Matrix {
+	return promql.Matrix{
+		{
+			Metric: labels.EmptyLabels(),
+			Floats: d.Samples,
+		},
+	}
+}
+
+func (q *Query) populateScalarFromScalarOperator(d types.ScalarData) promql.Scalar {
+	defer types.FPointSlicePool.Put(d.Samples, q.memoryConsumptionTracker)
```

Comment:
this seems like an odd place to put this. I mean it makes sense, we're done with the samples, but it feels easy for us to miss something and not return something to a pool somewhere.

Reply:
I'm not sure I follow sorry - we've consistently followed the pattern of "return the slice when we don't need the slice any more". This method is the place where we stop needing the slice, and after this point, we won't have a reference to the slice any more, so this is our last opportunity to return the slice to the pool.

Reply:
I was thinking as a defer on line 405 so that it's next to where we get the values.

Reply:
Are you talking about near line 388, where we call `GetValues()`? (line 405 is populating the warnings on the query results)

If so, it's not safe to return the slice at that point (unless it's in the `if` branch that calls `populateScalarFromScalarOperator`), as the slice will be used as-is for the matrix result.

Reply:
I think our line numbers might be getting confused. Ignoring that though, you're right, it would need to be in that if branch, so this seems like a good enough place :+1: